File: swift/Sources/Fory/UnknownCaseSerializer.swift
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import Foundation

public enum UnknownCaseSerializer {
    public static func writePayload(_ value: UnknownCase, _ context: WriteContext) throws {
        // Wire order is ref metadata first, then Any type metadata, then value bytes. Numeric
        // Any payloads are scalar values, so replay writes NotNullValue plus the original type id.
        if try writeTypedPayload(value, context) {
            return
        }
        try context.writeAny(value.value, refMode: .tracking, writeTypeInfo: true)
    }

    public static func readPayload(caseId: UInt32, _ context: ReadContext) throws -> UnknownCase {
        let rawFlag = try context.buffer.readInt8()
        guard let flag = RefFlag(rawValue: rawFlag) else {
            throw ForyError.refError("invalid ref flag \(rawFlag)")
        }
        switch flag {
        case .null:
            return UnknownCase(caseId: caseId, typeId: TypeId.unknown.rawValue, value: nil)
        case .ref:
            let refID = try context.buffer.readVarUInt32()
            let value = try context.refReader.readRefValue(refID)
            return UnknownCase(caseId: caseId, typeId: TypeId.unknown.rawValue, value: value)
        case .refValue:
            let reservedRefID = context.trackRef ? context.refReader.reserveRefID() : nil
            let (typeId, value) = try readNonNullPayload(context)
            if let reservedRefID {
                context.refReader.storeRef(value ?? NSNull(), at: reservedRefID)
            }
            return UnknownCase(caseId: caseId, typeId: typeId, value: value)
        case .notNullValue:
            let (typeId, value) = try readNonNullPayload(context)
            return UnknownCase(caseId: caseId, typeId: typeId, value: value)
        }
    }

    private static func writeTypedPayload(_ unknown: UnknownCase, _ context: WriteContext) throws -> Bool {
        guard let typeId = TypeId(rawValue: unknown.typeId), let value = unknown.value else {
            return false
        }
        switch typeId {
        case .bool:
            guard let typed = value as? Bool else { return false }
            writeRefAndType(typeId, context)
            context.buffer.writeUInt8(typed ? 1 : 0)
            return true
        case .int8:
            guard let typed = value as? Int8 else { return false }
            writeRefAndType(typeId, context)
            context.buffer.writeInt8(typed)
            return true
        case .uint8:
            guard let typed = value as? UInt8 else { return false }
            writeRefAndType(typeId, context)
            context.buffer.writeUInt8(typed)
            return true
        case .int16:
            guard let typed = value as? Int16 else { return false }
            writeRefAndType(typeId, context)
            context.buffer.writeInt16(typed)
            return true
        case .uint16:
            guard let typed = value as? UInt16 else { return false }
            writeRefAndType(typeId, context)
            context.buffer.writeUInt16(typed)
            return true
        case .int32:
            guard let typed = value as? Int32 else { return false }
            writeRefAndType(typeId, context)
            context.buffer.writeInt32(typed)
            return true
        case .varint32:
            guard let typed = value as? Int32 else { return false }
            writeRefAndType(typeId, context)
            context.buffer.writeVarInt32(typed)
            return true
        case .uint32:
            guard let typed = value as? UInt32 else { return false }
            writeRefAndType(typeId, context)
            context.buffer.writeUInt32(typed)
            return true
        case .varUInt32:
            guard let typed = value as? UInt32 else { return false }
            writeRefAndType(typeId, context)
            context.buffer.writeVarUInt32(typed)
            return true
        case .int64:
            guard let typed = value as? Int64 else { return false }
            writeRefAndType(typeId, context)
            context.buffer.writeInt64(typed)
            return true
        case .varint64:
            guard let typed = value as? Int64 else { return false }
            writeRefAndType(typeId, context)
            context.buffer.writeVarInt64(typed)
            return true
        case .taggedInt64:
            guard let typed = value as? Int64 else { return false }
            writeRefAndType(typeId, context)
            context.buffer.writeTaggedInt64(typed)
            return true
        case .uint64:
            guard let typed = value as? UInt64 else { return false }
            writeRefAndType(typeId, context)
            context.buffer.writeUInt64(typed)
            return true
        case .varUInt64:
            guard let typed = value as? UInt64 else { return false }
            writeRefAndType(typeId, context)
            context.buffer.writeVarUInt64(typed)
            return true
        case .taggedUInt64:
            guard let typed = value as? UInt64 else { return false }
            writeRefAndType(typeId, context)
            context.buffer.writeTaggedUInt64(typed)
            return true
        default:
            return false
        }
    }

    private static func writeRefAndType(_ typeId: TypeId, _ context: WriteContext) {
        context.buffer.writeInt8(RefFlag.notNullValue.rawValue)
        context.buffer.writeUInt8(UInt8(truncatingIfNeeded: typeId.rawValue))
    }

    private static func readNonNullPayload(_ context: ReadContext) throws -> (UInt32, Any?) {
        // UnknownCase owns the union payload envelope only. The envelope is not
        // a nested dynamic value, so depth checks belong to the decoded payload
        // serializer or the final root-context reset, not this carrier reader.
        let typeInfo = try context.readTypeInfo()
        let value = try context.readAnyValue(typeInfo: typeInfo)
        return (typeInfo.typeID.rawValue, value is ForyAnyNullValue ? nil : value)
    }
}

File: csharp/src/Fory/UnknownCaseSerializer.cs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

using System.Runtime.CompilerServices;

namespace Apache.Fory;

public static class UnknownCaseSerializer
{
    [MethodImpl(MethodImplOptions.NoInlining)]
    public static void WritePayload(WriteContext context, UnknownCase unknownCase)
    {
        // Wire order is ref metadata first, then Any type metadata, then value bytes. Numeric
        // Any payloads are scalar values, so replay writes NotNullValue plus the original type id.
        if (WriteTypedPayload(context, (TypeId)unknownCase.TypeId, unknownCase.Value))
        {
            return;
        }

        DynamicAnyCodec.WriteAny(context, unknownCase.Value, RefMode.Tracking, true, false);
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    public static UnknownCase ReadPayload(ReadContext context, int caseId)
    {
        RefFlag flag = context.RefReader.ReadRefFlag(context.Reader);
        switch (flag)
        {
            case RefFlag.Null:
                return UnknownCase.FromRuntime(caseId, (uint)TypeId.Unknown, null);
            case RefFlag.Ref:
                {
                    uint refId = context.RefReader.ReadRefId(context.Reader);
                    return UnknownCase.FromRuntime(caseId, (uint)TypeId.Unknown, context.RefReader.GetRefValue(refId));
                }
            case RefFlag.RefValue:
                {
                    uint reservedRefId = context.RefReader.ReserveRefId();
                    context.SetReservedRefId(reservedRefId);
                    try
                    {
                        (uint typeId, object? value) = ReadNonNullPayload(context);
                        context.StoreRef(value);
                        return UnknownCase.FromRuntime(caseId, typeId, value);
                    }
                    finally
                    {
                        context.ClearReservedRefId();
                    }
                }
            case RefFlag.NotNullValue:
                {
                    (uint typeId, object? value) = ReadNonNullPayload(context);
                    return UnknownCase.FromRuntime(caseId, typeId, value);
                }
            default:
                throw new RefException($"invalid ref flag {(sbyte)flag}");
        }
    }

    private static bool WriteTypedPayload(WriteContext context, TypeId typeId, object? value)
    {
        if (value is null)
        {
            return false;
        }

        switch (typeId)
        {
            case TypeId.Bool when value is bool typed:
                WriteRefAndType(context, typeId);
                context.Writer.WriteUInt8(typed ? (byte)1 : (byte)0);
                return true;
            case TypeId.Int8 when value is sbyte typed:
                WriteRefAndType(context, typeId);
                context.Writer.WriteInt8(typed);
                return true;
            case TypeId.UInt8 when value is byte typed:
                WriteRefAndType(context, typeId);
                context.Writer.WriteUInt8(typed);
                return true;
            case TypeId.Int16 when value is short typed:
                WriteRefAndType(context, typeId);
                context.Writer.WriteInt16(typed);
                return true;
            case TypeId.UInt16 when value is ushort typed:
                WriteRefAndType(context, typeId);
                context.Writer.WriteUInt16(typed);
                return true;
            case TypeId.Int32 when value is int typed:
                WriteRefAndType(context, typeId);
                context.Writer.WriteInt32(typed);
                return true;
            case TypeId.VarInt32 when value is int typed:
                WriteRefAndType(context, typeId);
                context.Writer.WriteVarInt32(typed);
                return true;
            case TypeId.UInt32 when value is uint typed:
                WriteRefAndType(context, typeId);
                context.Writer.WriteUInt32(typed);
                return true;
            case TypeId.VarUInt32 when value is uint typed:
                WriteRefAndType(context, typeId);
                context.Writer.WriteVarUInt32(typed);
                return true;
            case TypeId.Int64 when value is long typed:
                WriteRefAndType(context, typeId);
                context.Writer.WriteInt64(typed);
                return true;
            case TypeId.VarInt64 when value is long typed:
                WriteRefAndType(context, typeId);
                context.Writer.WriteVarInt64(typed);
                return true;
            case TypeId.TaggedInt64 when value is long typed:
                WriteRefAndType(context, typeId);
                context.Writer.WriteTaggedInt64(typed);
                return true;
            case TypeId.UInt64 when value is ulong typed:
                WriteRefAndType(context, typeId);
                context.Writer.WriteUInt64(typed);
                return true;
            case TypeId.VarUInt64 when value is ulong typed:
                WriteRefAndType(context, typeId);
                context.Writer.WriteVarUInt64(typed);
                return true;
            case TypeId.TaggedUInt64 when value is ulong typed:
                WriteRefAndType(context, typeId);
                context.Writer.WriteTaggedUInt64(typed);
                return true;
            default:
                return false;
        }
    }

    private static void WriteRefAndType(WriteContext context, TypeId typeId)
    {
        context.Writer.WriteInt8((sbyte)RefFlag.NotNullValue);
        context.Writer.WriteUInt8((byte)typeId);
    }

    private static (uint TypeId, object? Value) ReadNonNullPayload(ReadContext context)
    {
        // UnknownCase owns the union payload envelope only. The envelope is not
        // a nested dynamic value, so depth checks belong to the decoded payload
        // serializer or the final root-context reset, not this carrier reader.
        TypeInfo typeInfo = context.TypeResolver.ReadAnyTypeInfo(context);
        object? value = context.TypeResolver.ReadAnyValue(typeInfo, context);
        return ((uint)(typeInfo.WireTypeId ?? typeInfo.BuiltInTypeId ?? TypeId.Unknown), value);
    }
}
